File: clients/python/tests/test_client.py
import unittest
import io
import json
import base64
from zygote_harness import ZygoteClient, HarnessError

class TestZygoteClient(unittest.TestCase):
    def test_effect(self):
        # Prepare a mock stdin with a JSON-RPC response
        mock_res = {
            "jsonrpc": "2.0",
            "id": 1,
            "result": {
                "value": base64.b64encode(b"ReplayedAnswer").decode("utf-8")
            }
        }
        stdin = io.StringIO(json.dumps(mock_res) + "\n")
        stdout = io.StringIO()

        client = ZygoteClient(stdin=stdin, stdout=stdout)

        # Call effect
        res_bytes = client.effect("model", "prompt-1", b"MyPrompt")

        # Verify output
        self.assertEqual(res_bytes, b"ReplayedAnswer")

        # Verify what was sent to stdout
        sent = json.loads(stdout.getvalue())
        self.assertEqual(sent["method"], "effect")
        self.assertEqual(sent["params"]["op"], "model")
        self.assertEqual(sent["params"]["key"], "prompt-1")
        self.assertEqual(base64.b64decode(sent["params"]["value"]), b"MyPrompt")

    def test_error(self):
        mock_err = {
            "jsonrpc": "2.0",
            "id": 1,
            "error": {
                "code": -32600,
                "message": "divergence: effect mismatch"
            }
        }
        stdin = io.StringIO(json.dumps(mock_err) + "\n")
        stdout = io.StringIO()

        client = ZygoteClient(stdin=stdin, stdout=stdout)

        with self.assertRaises(HarnessError) as ctx:
            client.step("step-1")

        self.assertEqual(ctx.exception.code, -32600)
        self.assertIn("divergence", ctx.exception.message)

if __name__ == '__main__':
    unittest.main()

File: clients/typescript/tests/test_client.ts
import * as assert from 'assert';
import { ZygoteClient, HarnessError } from '../src/client';

async function runTests() {
    // We cannot easily intercept process.stdout synchronously in a simple Node test
    // without mocking `process.stdout.write`.
    const originalWrite = process.stdout.write.bind(process.stdout);

    let sentMsg: any = null;
    (process.stdout as any).write = (chunk: string | Uint8Array, cb?: any) => {
        try {
            sentMsg = JSON.parse(chunk.toString());
        } catch (e) {}
        return true;
    };

    const client = new ZygoteClient();

    // 1. Test effect
    const effectPromise = client.effect('model', 'prompt-1', new Uint8Array([104, 101, 108, 108, 111])); // "hello"

    // Simulate Zygote response on stdin
    process.stdin.emit('data', JSON.stringify({
        jsonrpc: '2.0',
        id: 1,
        result: { value: Buffer.from('world').toString('base64') }
    }) + '\n');

    const res = await effectPromise;
    assert.strictEqual(Buffer.from(res).toString(), 'world');
    assert.strictEqual(sentMsg.method, 'effect');
    assert.strictEqual(sentMsg.params.op, 'model');
    assert.strictEqual(sentMsg.params.key, 'prompt-1');
    assert.strictEqual(sentMsg.params.value, Buffer.from('hello').toString('base64'));

    // 2. Test error
    const stepPromise = client.step('step-1');

    process.stdin.emit('data', JSON.stringify({
        jsonrpc: '2.0',
        id: 2,
        error: { code: -32600, message: 'divergence: effect mismatch' }
    }) + '\n');

    try {
        await stepPromise;
        assert.fail('Expected HarnessError');
    } catch (e: any) {
        assert.strictEqual(e.code, -32600);
        assert.ok(e.message.includes('divergence'));
    }

    client.close();

    // Restore stdout
    process.stdout.write = originalWrite;
    console.log("All TS tests passed.");
}

runTests().catch(e => {
    console.error(e);
    process.exit(1);
});
